crates/ruff_server/src/session.rs

```diff
@@ -232,46 +236,55 @@ impl Workspaces {
 
     fn snapshot(&self, document_url: &Url) -> Option<DocumentRef> {
         self.workspace_for_url(document_url)
-            .and_then(|w| w.open_documents.snapshot(document_url))
+            .and_then(|(_, workspace)| workspace.open_documents.snapshot(document_url))
     }
 
     fn controller(&mut self, document_url: &Url) -> Option<&mut DocumentController> {
         self.workspace_for_url_mut(document_url)
-            .and_then(|w| w.open_documents.controller(document_url))
+            .and_then(|(_, workspace)| workspace.open_documents.controller(document_url))
     }
 
     fn configuration(&self, document_url: &Url) -> Option<&Arc<RuffConfiguration>> {
         self.workspace_for_url(document_url)
-            .map(|w| &w.configuration)
+            .map(|(_, workspace)| &workspace.configuration)
+    }
+
+    fn reload_configuration(&mut self, changed_url: &Url) -> crate::Result<()> {
+        let (path, workspace) = self
+            .workspace_for_url_mut(changed_url)
+            .ok_or_else(|| anyhow!("Workspace not found for {changed_url}"))?;
+        workspace.reload_configuration(path);
```

Comment:
How does configuration work here exactly? Do we find one configuration for the workspace, and then use that for all files?

Reply:
(I'm asking because I'd like to understand how it works if a user has subdirectories with distinct `ruff.toml` configurations that should only be applied to files in each respective subdirectory.)

Reply:
Good question! Right now, we only process one configuration file per workspace, but supporting this sub-folder configuration is on the roadmap. We do support this in a limited capacity at the moment - if a workspace is a subfolder of another workspace, we use the configuration file found in that workspace folder instead of the one in the parent workspace.

Reply:
How do you envision structuring / supporting it?

Reply:
We'd have a 'configuration index' in each workspace that would point to all valid configuration files and their locations. We could probably even cache the resolved configuration for this file path as part of the index. When taking a snapshot of a document, we'd find the corresponding resolved configuration from this index using the file path.